wagtail/admin/compare.py

```diff
@@ -55,8 +57,155 @@ def htmldiff(self):
+            comparison_class = get_comparison_class_for_block(block)
+
+            val_a = self.val_a[name]
+            val_b = self.val_b[name]
+
+            htmldiffs[name] = comparison_class(block, val_a, val_b).htmldiff()
+
+        return self.block.render_basic(htmldiffs)
+
+
+class StreamBlockComparison(BlockComparison):
+    def get_block_comparisons(self):
+        a_blocks = list(self.val_a) or []
+        b_blocks = list(self.val_b) or []
+
+        a_blocks_by_id = {block.id: block for block in a_blocks}
+        b_blocks_by_id = {block.id: block for block in b_blocks}
+
+        deleted_ids = a_blocks_by_id.keys() - b_blocks_by_id.keys()
+
+        comparisons = []
+        for block in b_blocks:
+            comparison_class = get_comparison_class_for_block(block.block)
+
+            if block.id in a_blocks_by_id:
+                # Changed/existing block
+                comparisons.append(comparison_class(block.block, a_blocks_by_id[block.id].value, block.value))
+            else:
+                # New block
+                comparisons.append(comparison_class(block.block, None, block.value))
+
+        # Insert deleted blocks at the index where they used to be
+        deleted_block_indices = [(block, i) for i, block in enumerate(a_blocks) if block.id in deleted_ids]
+
+        for block, index in deleted_block_indices:
+            comparison_class = get_comparison_class_for_block(block.block)
+
+            # Insert the block back in where it was before it was deleted.
+            # Note: we need to account for new blocks when finding the position.
+            current_index = 0
+            block_inserted = False
+            for i, comparison in enumerate(comparisons):
+                if comparison.is_new():
+                    continue
+
+                if current_index == index:
+                    comparisons.insert(i, comparison_class(block.block, block.value, None))
+                    block_inserted = True
+                    break
+
+                current_index += 1
+
+            # Deleted block was from the end
+            if not block_inserted:
+                comparisons.append(comparison_class(block.block, block.value, None))
+
+        return comparisons
+
+    def htmldiff(self):
+        comparisons_html = []
+
+        for comparison in self.get_block_comparisons():
+            classes = ['comparison__child-object']
+            if comparison.is_new():
+                classes.append('addition')
+                block_rendered = comparison.block.render_basic(comparison.val_b)
+            elif comparison.is_deleted():
+                classes.append('deletion')
+                block_rendered = comparison.block.render_basic(comparison.val_a)
+            elif comparison.has_changed():
+                block_rendered = comparison.htmldiff()
+            else:
+                block_rendered = comparison.block.render_basic(comparison.val_a)
```

Comment:
The use of `comparison.block.render_basic` for everything except the `has_changed` case means that comparison class `htmldiff` methods can't provide markup for the addition/deletion/same cases.

For example, [this logic](https://github.com/wagtail/wagtail/blob/5e5484d4d7d3fec0b860826a07903165ed6dfb31/wagtail/images/templates/wagtailimages/widgets/compare.html#L15) in the image comparison template (used [here](https://github.com/wagtail/wagtail/pull/4933/files#diff-70247e933fb4da7e546dcaf823810235R36) in this PR) doesn't ever get called, since the template isn't rendered in those cases. Instead, the `block.render_basic` rendering is used, which just renders the image at its original size. This isn't great in cases where the image is very large.

When an image is changed, the nice diff view is shown:

![image](https://user-images.githubusercontent.com/654645/50706515-19dc8a80-102c-11e9-8805-1f1fdffb697a.png)

But when the image is added or deleted, it's displayed at its original size:

![image](https://user-images.githubusercontent.com/654645/50706559-31b40e80-102c-11e9-8acd-4032d798f78e.png)

Could the `htmldiff` methods _always_ generate the rendering to be used? Even when there's no difference? They'd just render the current state of the block. Perhaps the base `BlockComparison` could provide some kind of fallback that inherited comparison classes use?